pulp_certguard/app/models.py

```diff
@@ -0,0 +1,162 @@
+        body = body.replace(footer, '')
+        body = body.strip(' \n\r')
+        return '\n'.join((header, body, footer))
+
+    @staticmethod
+    def load(pem):
+        """
+        Load the PEM encoded certificate.
+
+        Encapsulates complexity of OpenSSL.
+
+        Args:
+            pem (str): A PEM encoded certificate.
+
+        Returns:
+            openssl.X509: The loaded certificate.
+
+        Raises:
+            ValueError: On load failed.
+        """
+        try:
+            return openssl.load_certificate(openssl.FILETYPE_PEM, buffer=Validator.format(pem))
+        except Exception as le:
+            raise ValueError(str(le))
+
+    @staticmethod
+    def client_certificate(request):
+        """
+        Extract and load the client certificate passed in the SSL-CLIENT-CERTIFICATE header.
+
+        Args:
+            request (aiohttp.web.Request): A request for a published file.
+
+        Returns:
+            openssl.X509: The loaded certificate.
+
+        Raises:
+            KeyError: When the client certificate header has not
+                been passed in the request.
+        """
+        name = 'SSL-CLIENT-CERTIFICATE'
+        try:
+            certificate = request.headers[name]
+        except KeyError:
+            reason = _(
+                'HTTP header "{h}" not found for: {u}.').format(
+                    h=name,
+                    u=request.url)
+            log.debug(reason)
+            raise KeyError(reason)
+        else:
+            return Validator.load(certificate)
+
+    def __init__(self, ca_certificate):
+        """
+        Args:
+            ca_certificate (str): A PEM encoded CA certificate.
+        """
+        self.ca_certificate = self.load(ca_certificate)
+
+    @property
+    def store(self):
+        """
+        A X509 certificate (trust) store.
+
+        Returns:
+            openssl.X509Store: A store containing the CA certificate.
+        """
+        store = openssl.X509Store()
+        store.add_cert(self.ca_certificate)
+        return store
+
+    def __call__(self, request):
+        """
+        Validate the client X.509 certificate passed in the request.
+
+        Args:
+            request (aiohttp.web.Request): A request for a published file.
+
+        Raises:
+            PermissionError: When validation the client certificate
+                cannot be validated.
+        """
+        try:
+            context = openssl.X509StoreContext(
+                certificate=self.client_certificate(request),
+                store=self.store,
+            )
+            context.verify_certificate()
+        except openssl.X509StoreContextError:
+            reason = _('Client certificate cannot be validated.')
+            log.debug(reason)
+            raise PermissionError(reason)
+        except Exception as pe:
+            reason = str(pe)
+            log.debug(reason, exc_info=True)
+            raise PermissionError(reason)
```

Comment:
I think this is going to be called from the aiohttp handler. As is I think it would return a http 500 to the user, but I think it should be a 403? aiohttp has that exception type and maybe we can raise it here?

Reply:
similarly for above and through docs.

Reply:
Agreed, it should result in a 403.
The design already documented in core for content guards is for the `ContentGuard.permit()` to raise `PermissionError` which is expected to be caught by the content app which returns the 403.  This is how the content app worked before consolidation.  I'm fine with raising an aiohttp `Forbidden` exception instead.  I mainly wanted it to be an existing generic permission related exception.

Can you link me to docs on the aiohttp exception?

Reply:
I found `aiohttp.web_exceptions.HTTPForbidden` but nothing in the documentation about it or how it's used.  This what you had in mind?